Add type NDEFRecordType to web-nfc.d.ts

drewmw5 opened the pull request:
This pull request is for adding types to recordTypes.
- It adds a new type: "NDEFRecordTypes"
- It includes all valid record types for recordTypes referenced from https://developer.mozilla.org/en-US/docs/Web/API/NDEFReader/write
- Values consist of: 
   "absolute-url"
   "empty"
   "mime"
   "smart-poster"
   "text"
   "unknown"
   "URL"

drewmw5 committed:
Add type NDEFRecordType to web-nfc.d.ts

This pull request is for adding types to recordTypes.
- It adds "type NDEFRecordTypes"
- It includes all valid record types referenced from https://developer.mozilla.org/en-US/docs/Web/API/NDEFReader/write
- Values consist of: 
   "absolute-url"
   "empty"
   "mime"
   "smart-poster"
   "text"
   "unknown"
   "URL"

beaufortfrancois commented:
As you can see at https://w3c.github.io/web-nfc/#data-mapping, `recordType` can be any string that follows a certain pattern (See local type and external type).

Is it possible in TypeScript to have something like below?
```ts
  readonly recordType: NDEFRecordTypes | string;
```
or 
```ts
type NDEFRecordTypes = string | "absolute-url" | "empty" | "mime" | "smart-poster" | "text" | "unknown" | "URL";
```

drewmw5 commented:
> As you can see at https://w3c.github.io/web-nfc/#data-mapping, `recordType` can be any string that follows a certain pattern (See local type and external type).
> 
> Is it possible in TypeScript to have something like below?
> 
> ```ts
>   readonly recordType: NDEFRecordTypes | string;
> ```
> 
> or
> 
> ```ts
> type NDEFRecordTypes = string | "absolute-url" | "empty" | "mime" | "smart-poster" | "text" | "unknown" | "URL";
> ```

Ah, good to know, there's still plenty I have left to read. Thanks for that. 
You are correct, the following example you gave is valid TypeScript, the other one was not:

```ts
type NDEFRecordTypes = string | "absolute-url" | "empty" | "mime" | "smart-poster" | "text" | "unknown" | "URL";
```
Is there more to it than that?

drewmw5 committed:
Added "string" type to NDEFRecordtypes

Added additional string type to NDEFRecordTypes as suggested by @beaufortfrancois

drewmw5 commented:
> > As you can see at https://w3c.github.io/web-nfc/#data-mapping, `recordType` can be any string that follows a certain pattern (See local type and external type).
> > Is it possible in TypeScript to have something like below?
> > ```ts
> >   readonly recordType: NDEFRecordTypes | string;
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > or
> > ```ts
> > type NDEFRecordTypes = string | "absolute-url" | "empty" | "mime" | "smart-poster" | "text" | "unknown" | "URL";
> > ```
> 
> Ah, good to know, there's still plenty I have left to read. Thanks for that. You are correct, the following example you gave is valid TypeScript, the other one was not:
> 
> ```ts
> type NDEFRecordTypes = string | "absolute-url" | "empty" | "mime" | "smart-poster" | "text" | "unknown" | "URL";
> ```
> 
> I've gone ahead and updated the original commit. Is there more to it than that?

beaufortfrancois commented on web-nfc.d.ts:
```diff
@@ -22,9 +22,11 @@ declare type NDEFRecordDataSource = string | BufferSource | NDEFMessageInit
 interface Window {
   NDEFRecord: NDEFRecord
 }
+type NDEFRecordTypes = string | "absolute-url" | "empty" | "mime" | "smart-poster" | "text" | "unknown" | "URL";
```
Should we have `declare` as well there?

takefumi-yoshii replied:
It is incorrect to include `string` here. In this definition, it is synonymous with `string`.
If exclude `string`, we will get the expected compilation error. [What this means](https://www.typescriptlang.org/play?#code/C4TwDgpgBA+gcgEQKIDEBKEDGB7ATgEwBVwIBnKAXilOFwEsA7AcygB8oAiAQwCNTsANgFdgEALRDcAjm04QAtmFAz2HeXXkQVnUvK65gYsNhoRc2jqIAewC0IYBrBtgDuDCwFU0AGQ4BuAFgAKGCAelCoQgALM2g6cmcoM1w8YJwGGmouTQBBUgBlWkYmAC5YRFQMHAJiSHIqAHJ1Gi4HCAbg4NBIKAr0LDwiEnrOXn5hUQkpCwUlEAt1TQtdfUNjU3NZSwgbO0dnN08ff06g8MiY3DjyaqvMYCTcFNw07AyHnch7iHwkJ7wyn0qoNamRKFAmvFgK12sEgA)

Union types of such `string literal type` should be used to avoid including incorrect strings.

zolkis replied:
Yes, if we want a language level (early) error. Just noting that it's also possible to defer that decision (to accept or not a given string record type) to implementations: then a general string should be fine, including `"mistake"`. It depends how do we want to do it.
IMHO we (Web NFC) are not in the position to extend the range of valid NFC record types or leave extension possibilities to implementations, therefore I agree on using string literals.

takefumi-yoshii replied:
It is correct that "NDEF Record Type" can specify any character string (local type and external type), So it should be a `string`.Committing as a document is good, but it doesn't contribute to Type Safe.

drewmw5 replied:
I just went ahead and updated it to include the declare keyword and also removed the string type.

beaufortfrancois commented on web-nfc.d.ts:
```diff
@@ -22,9 +22,11 @@ declare type NDEFRecordDataSource = string | BufferSource | NDEFMessageInit
 interface Window {
   NDEFRecord: NDEFRecord
 }
+type NDEFRecordTypes = string | "absolute-url" | "empty" | "mime" | "smart-poster" | "text" | "unknown" | "URL";
```
I believe it should be "url", not "URL".

takefumi-yoshii replied:
plus one 👍

drewmw5 replied:
Good catch. Fixed

beaufortfrancois commented:
@takefumi-yoshii Can you have a look as well?

drewmw5 committed:
Changed "URL" to "url"

drewmw5 committed:
Added declare keyword; removed string type from NDEFReaderTypes

zolkis commented:
As @beaufortfrancois mentioned [earlier](https://github.com/w3c/web-nfc/pull/629#issuecomment-972637525), record type can be a string. How would you write this example after this PR?
https://w3c.github.io/web-nfc/#write-an-external-record-with-an-ndef-message-as-payload 

Then, smart poster local types should also be listed?

zolkis commented:
Well, to me it seems that if we cannot add to the definitions a generic "string" as a fallback, then we should leave `recordType` as `string`. Static type validation as "string" should be enough, and dynamic validation, including standard record types defined by the NFC Forum, is defined by the algorithms in the spec, and it will depend on context (where the records belong). 
If there are shortcomings there, please file issues.
